eth/downloader/skeleton.go

```diff
@@ -223,53 +232,63 @@ func (s *skeleton) startup() {
 	// sync loop was torn down for good.
 	defer close(s.terminated)
 
-	// Wait for startup or teardown
-	select {
-	case errc := <-s.terminate:
-		// No head was announced but Geth is shutting down
-		errc <- nil
-		return
+	// Wait for startup or teardown. This wait might loop a few times if a beacon
+	// client requests sync head extensions, but not forced reorgs (i.e. they are
+	// giving us new payloads without setting a starting head intially).
+	for {
+		select {
+		case errc := <-s.terminate:
+			// No head was announced but Geth is shutting down
+			errc <- nil
+			return
 
-	case head := <-s.headEvents:
-		// New head announced, start syncing to it, looping every time a current
-		// cycle is terminated due to a chain event (head reorg, old chain merge)
-		s.started = time.Now()
+		case event := <-s.headEvents:
+			// New head announced, start syncing to it, looping every time a current
+			// cycle is terminated due to a chain event (head reorg, old chain merge).
+			if !event.force {
+				event.errc <- errors.New("forced head needed for startup")
+				continue
+			}
+			event.errc <- nil // forced head accepted for startup
+			head := event.header
+			s.started = time.Now()
+
+			for {
+				// If the sync cycle terminated or was terminated, propagate up when
+				// higher layers request termination. There's no fancy explicit error
+				// signalling as the sync loop should never terminate (TM).
+				newhead, err := s.sync(head)
+				switch {
+				case err == errSyncLinked:
+					// Sync cycle linked up to the genesis block. Tear down the loop
+					// and restart it so, it can properly notify the backfiller. Don't
+					// account a new head.
+					head = nil
+
+				case err == errSyncMerged:
+					// Subchains were merged, we just need to reinit the internal
+					// start to continue on the tail of the merged chain. Don't
+					// announce a new head,
+					head = nil
+
+				case err == errSyncReorged:
+					// The subchain being synced got modified at the head in a
+					// way that requires resyncing it. Restart sync with the new
+					// head to force a cleanup.
+					head = newhead
+
+				case err == errTerminated:
+					// Sync was requested to be terminated from within, stop and
+					// return (no need to pass a message, was already done internally)
+					return
```

Comment:
Shouldn't we close the errc here? Otherwise the caller might wait forever?

Reply:
There's no errc here. The place where this error originates from handlers errc locally.

Reply:
L407